src/com/googlecode/jmxtrans/model/output/GraphiteWriter.java

```diff
@@ -101,6 +101,11 @@ public void internalWrite(Server server, Query query, ImmutableList<Result> resu
 							log.debug("Graphite Message: {}", line);
 							writer.write(line);
 							writer.flush();
+							if (writer.checkError()) {
+								log.error("Error writing to Graphite, clearing Graphite socket pool");
+								pool.returnObject(address, socket);
+								pool.clear();
```

Comment:
I think the idea is it's better to be overzealous and clear the entire pool, since it incurs minimal overhead and ensures that future writes will succeed. If one persistent connection is dead the other connections have suffered through the same issue and there is a good chance they are dead as well.

As for the thread safety, if another connection is checked out and the write succeeds then it's all good, if it fails then it will trigger the connection reset again and be good to go on the next write.

In a pathological case where every connection is checked out we would lose as many datagrams as there are connections, but even this seems not too bad compared to further complicating the code.